From 3bf521a11e76879353086d19d62f71d18919a3ca Mon Sep 17 00:00:00 2001
From: Matthias Diener <mdiener@illinois.edu>
Date: Mon, 31 Jan 2022 05:22:35 +0100
Subject: [PATCH] Fix value of CL_DEVICE_PCI_DOMAIN_ID_NV

---
 include/CL/cl_ext.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/CL/cl_ext.h b/include/CL/cl_ext.h
index 864eca488..7007f6cf3 100644
--- a/include/CL/cl_ext.h
+++ b/include/CL/cl_ext.h
@@ -230,7 +230,7 @@ typedef CL_API_ENTRY cl_command_queue
 #define CL_DEVICE_PCI_SLOT_ID_NV                    0x4009
 
 /* extension to cl_nv_device_attribute_query */
-#define CL_DEVICE_PCI_DOMAIN_ID_NV                  0x4010
+#define CL_DEVICE_PCI_DOMAIN_ID_NV                  0x400A
 
 /*********************************
 * cl_amd_device_attribute_query *
