diff --git a/lib/CL/clGetPlatformIDs.c b/lib/CL/clGetPlatformIDs.c
index e5c48b56..facae1ee 100644
--- a/lib/CL/clGetPlatformIDs.c
+++ b/lib/CL/clGetPlatformIDs.c
@@ -37,115 +37,115 @@
 */
 #ifdef BUILD_ICD
 struct _cl_icd_dispatch pocl_dispatch = {
-  &POclGetPlatformIDs,
-  &POclGetPlatformInfo,
-  &POclGetDeviceIDs,
-  &POclGetDeviceInfo,
-  &POclCreateContext,
-  &POclCreateContextFromType,
-  &POclRetainContext,
-  &POclReleaseContext,
-  &POclGetContextInfo,
-  &POclCreateCommandQueue,
-  &POclRetainCommandQueue, /* 10 */
-  &POclReleaseCommandQueue,
-  &POclGetCommandQueueInfo,
+  &POname(clGetPlatformIDs),
+  &POname(clGetPlatformInfo),
+  &POname(clGetDeviceIDs),
+  &POname(clGetDeviceInfo),
+  &POname(clCreateContext),
+  &POname(clCreateContextFromType),
+  &POname(clRetainContext),
+  &POname(clReleaseContext),
+  &POname(clGetContextInfo),
+  &POname(clCreateCommandQueue),
+  &POname(clRetainCommandQueue), /* 10 */
+  &POname(clReleaseCommandQueue),
+  &POname(clGetCommandQueueInfo),
   NULL /*clSetCommandQueueProperty*/,
-  &POclCreateBuffer,
-  &POclCreateImage2D,
-  &POclCreateImage3D,
-  &POclRetainMemObject,
-  &POclReleaseMemObject,
-  &POclGetSupportedImageFormats,
-  &POclGetMemObjectInfo, /* 20 */
-  &POclGetImageInfo,
-  &POclCreateSampler,
-  &POclRetainSampler,
-  &POclReleaseSampler,
-  &POclGetSamplerInfo,
-  &POclCreateProgramWithSource,
-  &POclCreateProgramWithBinary,
-  &POclRetainProgram,
-  &POclReleaseProgram,
-  &POclBuildProgram, /* 30 */
-  &POclUnloadCompiler,
-  &POclGetProgramInfo,
-  &POclGetProgramBuildInfo,
-  &POclCreateKernel,
-  &POclCreateKernelsInProgram,
-  &POclRetainKernel,
-  &POclReleaseKernel,
-  &POclSetKernelArg,
-  &POclGetKernelInfo,
-  &POclGetKernelWorkGroupInfo, /* 40 */
-  &POclWaitForEvents,
-  &POclGetEventInfo,
-  &POclRetainEvent,
-  &POclReleaseEvent,
-  &POclGetEventProfilingInfo,
-  &POclFlush,
-  &POclFinish,
-  &POclEnqueueReadBuffer,
-  &POclEnqueueWriteBuffer,
-  &POclEnqueueCopyBuffer, /* 50 */
-  &POclEnqueueReadImage,
-  &POclEnqueueWriteImage,
-  &POclEnqueueCopyImage,
-  &POclEnqueueCopyImageToBuffer,
-  &POclEnqueueCopyBufferToImage,
-  &POclEnqueueMapBuffer,
-  &POclEnqueueMapImage,
-  &POclEnqueueUnmapMemObject,
-  &POclEnqueueNDRangeKernel,
-  &POclEnqueueTask, /* 60 */
-  &POclEnqueueNativeKernel,
-  &POclEnqueueMarker,
-  &POclEnqueueWaitForEvents,
-  &POclEnqueueBarrier,
-  &POclGetExtensionFunctionAddress,
-  &POclCreateFromGLBuffer,
-  &POclCreateFromGLTexture2D,
-  &POclCreateFromGLTexture3D,
-  &POclCreateFromGLRenderbuffer,
-  &POclGetGLObjectInfo,
-  &POclGetGLTextureInfo,
-  &POclEnqueueAcquireGLObjects,
-  &POclEnqueueReleaseGLObjects,
-  &POclGetGLContextInfoKHR,
+  &POname(clCreateBuffer),
+  &POname(clCreateImage2D),
+  &POname(clCreateImage3D),
+  &POname(clRetainMemObject),
+  &POname(clReleaseMemObject),
+  &POname(clGetSupportedImageFormats),
+  &POname(clGetMemObjectInfo), /* 20 */
+  &POname(clGetImageInfo),
+  &POname(clCreateSampler),
+  &POname(clRetainSampler),
+  &POname(clReleaseSampler),
+  &POname(clGetSamplerInfo),
+  &POname(clCreateProgramWithSource),
+  &POname(clCreateProgramWithBinary),
+  &POname(clRetainProgram),
+  &POname(clReleaseProgram),
+  &POname(clBuildProgram), /* 30 */
+  &POname(clUnloadCompiler),
+  &POname(clGetProgramInfo),
+  &POname(clGetProgramBuildInfo),
+  &POname(clCreateKernel),
+  &POname(clCreateKernelsInProgram),
+  &POname(clRetainKernel),
+  &POname(clReleaseKernel),
+  &POname(clSetKernelArg),
+  &POname(clGetKernelInfo),
+  &POname(clGetKernelWorkGroupInfo), /* 40 */
+  &POname(clWaitForEvents),
+  &POname(clGetEventInfo),
+  &POname(clRetainEvent),
+  &POname(clReleaseEvent),
+  &POname(clGetEventProfilingInfo),
+  &POname(clFlush),
+  &POname(clFinish),
+  &POname(clEnqueueReadBuffer),
+  &POname(clEnqueueWriteBuffer),
+  &POname(clEnqueueCopyBuffer), /* 50 */
+  &POname(clEnqueueReadImage),
+  &POname(clEnqueueWriteImage),
+  &POname(clEnqueueCopyImage),
+  &POname(clEnqueueCopyImageToBuffer),
+  &POname(clEnqueueCopyBufferToImage),
+  &POname(clEnqueueMapBuffer),
+  &POname(clEnqueueMapImage),
+  &POname(clEnqueueUnmapMemObject),
+  &POname(clEnqueueNDRangeKernel),
+  &POname(clEnqueueTask), /* 60 */
+  &POname(clEnqueueNativeKernel),
+  &POname(clEnqueueMarker),
+  &POname(clEnqueueWaitForEvents),
+  &POname(clEnqueueBarrier),
+  &POname(clGetExtensionFunctionAddress),
+  &POname(clCreateFromGLBuffer),
+  &POname(clCreateFromGLTexture2D),
+  &POname(clCreateFromGLTexture3D),
+  &POname(clCreateFromGLRenderbuffer),
+  &POname(clGetGLObjectInfo),
+  &POname(clGetGLTextureInfo),
+  &POname(clEnqueueAcquireGLObjects),
+  &POname(clEnqueueReleaseGLObjects),
+  &POname(clGetGLContextInfoKHR),
   NULL, /* &clUnknown75 */
   NULL, /* &clUnknown76 */
   NULL, /* &clUnknown77 */
   NULL, /* &clUnknown78 */
   NULL, /* &clUnknown79 */
   NULL, /* &clUnknown80 */
-  &POclSetEventCallback,
-  &POclCreateSubBuffer,
-  &POclSetMemObjectDestructorCallback,
-  &POclCreateUserEvent,
-  &POclSetUserEventStatus,
-  &POclEnqueueReadBufferRect,
-  &POclEnqueueWriteBufferRect,
-  &POclEnqueueCopyBufferRect,
-  NULL, /* &POclCreateSubDevicesEXT,     */
-  &POclRetainDevice, /* &POclRetainDeviceEXT,         */
-  &POclReleaseDevice, /* &POclReleaseDeviceEXT,        */
+  &POname(clSetEventCallback),
+  &POname(clCreateSubBuffer),
+  &POname(clSetMemObjectDestructorCallback),
+  &POname(clCreateUserEvent),
+  &POname(clSetUserEventStatus),
+  &POname(clEnqueueReadBufferRect),
+  &POname(clEnqueueWriteBufferRect),
+  &POname(clEnqueueCopyBufferRect),
+  NULL, /* &POname(clCreateSubDevicesEXT),     */
+  &POname(clRetainDevice), /* &POname(clRetainDeviceEXT),         */
+  &POname(clReleaseDevice), /* &POname(clReleaseDeviceEXT),        */
   NULL, /* &clUnknown92 */
-  &POclCreateSubDevices,
-  &POclRetainDevice,
-  &POclReleaseDevice,
-  &POclCreateImage,
-  &POclCreateProgramWithBuiltInKernels,
-  &POclCompileProgram,
-  &POclLinkProgram,
-  &POclUnloadPlatformCompiler, 
-  &POclGetKernelArgInfo,
-  &POclEnqueueFillBuffer,
-  &POclEnqueueFillImage,
-  &POclEnqueueMigrateMemObjects,
-  &POclEnqueueMarkerWithWaitList,
-  &POclEnqueueBarrierWithWaitList,
-  &POclGetExtensionFunctionAddressForPlatform,
-  &POclCreateFromGLTexture,
+  &POname(clCreateSubDevices),
+  &POname(clRetainDevice),
+  &POname(clReleaseDevice),
+  &POname(clCreateImage),
+  &POname(clCreateProgramWithBuiltInKernels),
+  &POname(clCompileProgram),
+  &POname(clLinkProgram),
+  &POname(clUnloadPlatformCompiler), 
+  &POname(clGetKernelArgInfo),
+  &POname(clEnqueueFillBuffer),
+  &POname(clEnqueueFillImage),
+  &POname(clEnqueueMigrateMemObjects),
+  &POname(clEnqueueMarkerWithWaitList),
+  &POname(clEnqueueBarrierWithWaitList),
+  &POname(clGetExtensionFunctionAddressForPlatform),
+  &POname(clCreateFromGLTexture),
   NULL, /* &clUnknown109 */
   NULL, /* &clUnknown110 */
   NULL, /* &clUnknown111 */
@@ -161,19 +161,19 @@ struct _cl_icd_dispatch pocl_dispatch = {
   NULL, /* &clUnknown121 */
   NULL, /* &clUnknown122 */
 #if (OCL_ICD_IDENTIFIED_FUNCTIONS > 110)
-  &POclCreateCommandQueueWithProperties,
-  NULL, /* &POclCreatePipe,*/
-  NULL, /* &POclGetPipeInfo,*/
-  &POclSVMAlloc,
-  &POclSVMFree,
-  &POclEnqueueSVMFree,
-  &POclEnqueueSVMMemcpy,
-  &POclEnqueueSVMMemFill,
-  &POclEnqueueSVMMap,
-  &POclEnqueueSVMUnmap,
+  &POname(clCreateCommandQueueWithProperties),
+  NULL, /* &POname(clCreatePipe),*/
+  NULL, /* &POname(clGetPipeInfo),*/
+  &POname(clSVMAlloc),
+  &POname(clSVMFree),
+  &POname(clEnqueueSVMFree),
+  &POname(clEnqueueSVMMemcpy),
+  &POname(clEnqueueSVMMemFill),
+  &POname(clEnqueueSVMMap),
+  &POname(clEnqueueSVMUnmap),
   NULL, /* clCreateSamplerWithProperties */
-  &POclSetKernelArgSVMPointer,
-  &POclSetKernelExecInfo,
+  &POname(clSetKernelArgSVMPointer),
+  &POname(clSetKernelExecInfo),
   NULL, /* &clUnknown136 */
   NULL, /* &clUnknown137 */
   NULL, /* &clUnknown138 */
diff --git a/lib/CL/pocl_cl.h b/lib/CL/pocl_cl.h
index fbe8dd25..23d8546c 100644
--- a/lib/CL/pocl_cl.h
+++ b/lib/CL/pocl_cl.h
@@ -210,9 +210,6 @@ typedef pthread_mutex_t pocl_lock_t;
 #ifdef __APPLE__
 /* Note: OSX doesn't support aliases because it doesn't use ELF */
 
-#  ifdef BUILD_ICD
-#    error "ICD not supported on OSX"
-#  endif
 #  define POname(name) name
 #  define POdeclsym(name)
 #  define POsym(name)
@@ -229,17 +226,10 @@ typedef pthread_mutex_t pocl_lock_t;
 #else
 /* Symbol aliases are supported */
 
-#  define POname(name) PO##name
-#  define POdeclsym(name)                      \
-  __typeof__(name) PO##name __attribute__((visibility("hidden")));
-#  define POCL_ALIAS_OPENCL_SYMBOL(name)                                \
-  __typeof__(name) name __attribute__((alias ("PO" #name), visibility("default")));
-#  define POsymAlways(name) POCL_ALIAS_OPENCL_SYMBOL(name)
-#  if !defined(BUILD_ICD)
-#    define POsym(name) POCL_ALIAS_OPENCL_SYMBOL(name)
-#  else
-#    define POsym(name)
-#  endif
+#  define POname(name) name
+#  define POdeclsym(name)
+#  define POsym(name)
+#  define POsymAlways(name)
 
 #endif
 
