diff --git a/lib/CL/CMakeLists.txt b/lib/CL/CMakeLists.txt
index 4c016d3f..098c9388 100644
--- a/lib/CL/CMakeLists.txt
+++ b/lib/CL/CMakeLists.txt
@@ -133,6 +133,7 @@ set(POCL_LIB_SOURCES  "clCreateContextFromType.c"
                    "pocl_runtime_config.c" "pocl_runtime_config.h"
                    "pocl_mem_management.c"  "pocl_mem_management.h"
                    "pocl_hash.c"
+                   "pocl_paths.cc"
                    "pocl_debug.h" "pocl_debug.c" "pocl_timing.c"
                    "clSVMAlloc.c" "clSVMFree.c" "clEnqueueSVMFree.c"
                    "clEnqueueSVMMap.c" "clEnqueueSVMUnmap.c"
diff --git a/lib/CL/pocl_llvm_build.cc b/lib/CL/pocl_llvm_build.cc
index a6174415..0dc835ff 100644
--- a/lib/CL/pocl_llvm_build.cc
+++ b/lib/CL/pocl_llvm_build.cc
@@ -72,6 +72,12 @@ POP_COMPILER_DIAGS
 /* Global pocl device to be used by passes if needed */
 cl_device_id currentPoclDevice = NULL;
 
+extern "C"
+{
+  const char *conda_pocl_get_clang_resource_dir();
+  const char *conda_pocl_get_pocl_install_private_datadir();
+}
+
 
 //#define DEBUG_POCL_LLVM_API
 
@@ -419,16 +425,21 @@ int pocl_llvm_build_program(cl_program program,
   if (0) {
 #endif
   } else {
-    kernelh = POCL_INSTALL_PRIVATE_DATADIR;
+    kernelh = conda_pocl_get_pocl_install_private_datadir();
   }
   BuiltinRenamesH = kernelh;
   kernelh += "/include/_kernel.h";
   BuiltinRenamesH += "/include/_builtin_renames.h";
 
+
   po.Includes.push_back(BuiltinRenamesH);
 #ifndef LLVM_OLDER_THAN_4_0
+  std::string clang_header_dir =
+      std::string(conda_pocl_get_clang_resource_dir())
+      + "/include/opencl-c.h";
+
   // Use Clang's opencl-c.h header.
-  po.Includes.push_back(CLANG_RESOURCE_DIR "/include/opencl-c.h");
+  po.Includes.push_back(clang_header_dir);
 #endif
   po.Includes.push_back(kernelh);
   clang::TargetOptions &ta = pocl_build.getTargetOpts();
@@ -832,7 +843,7 @@ kernel_library
     kernellib += subdir;
   } else // POCL_BUILDING == 0, use install dir
 #endif
-  kernellib = POCL_INSTALL_PRIVATE_DATADIR;
+  kernellib = conda_pocl_get_pocl_install_private_datadir();
   kernellib += "/kernel-";
   kernellib += device->llvm_target_triplet;
   if (is_host) {
diff --git a/lib/CL/pocl_paths.cc b/lib/CL/pocl_paths.cc
new file mode 100644
index 00000000..8269c034
--- /dev/null
+++ b/lib/CL/pocl_paths.cc
@@ -0,0 +1,14 @@
+#include "config.h"
+
+extern "C"
+{
+  const char *conda_pocl_get_clang_resource_dir()
+  {
+    return CLANG_RESOURCE_DIR;
+  }
+
+  const char *conda_pocl_get_pocl_install_private_datadir()
+  {
+    return POCL_INSTALL_PRIVATE_DATADIR;
+  }
+}
