diff --git a/lib/CL/pocl_llvm_build.cc b/lib/CL/pocl_llvm_build.cc
index 6b5c8168..024124ad 100644
--- a/lib/CL/pocl_llvm_build.cc
+++ b/lib/CL/pocl_llvm_build.cc
@@ -432,7 +432,7 @@ int pocl_llvm_build_program(cl_program program,
   la->setStackProtector(LangOptions::StackProtectorMode::SSPOff);
 
   la->PICLevel = PICLevel::BigPIC;
-  la->PIE = 1;
+  la->PIE = 0;
 
   std::string IncludeRoot;
   std::string KernelH;
@@ -573,8 +573,8 @@ int pocl_llvm_build_program(cl_program program,
 
   if ((*mod)->getModuleFlag("PIC Level") == nullptr)
     (*mod)->setPICLevel(PICLevel::BigPIC);
-  if ((*mod)->getModuleFlag("PIE Level") == nullptr)
-    (*mod)->setPIELevel(PIELevel::Large);
+  //if ((*mod)->getModuleFlag("PIE Level") == nullptr)
+  //  (*mod)->setPIELevel(PIELevel::Large);
 
   // link w kernel lib, but not if we're called from clCompileProgram()
   // Later this should be replaced with indexed linking of source code
@@ -668,8 +668,8 @@ int pocl_llvm_link_program(cl_program program, unsigned device_i,
     linked_module->setDataLayout(libmodule->getDataLayout());
     if (linked_module->getModuleFlag("PIC Level") == nullptr)
       linked_module->setPICLevel(PICLevel::BigPIC);
-    if (linked_module->getModuleFlag("PIE Level") == nullptr)
-      linked_module->setPIELevel(PIELevel::Large);
+    //if (linked_module->getModuleFlag("PIE Level") == nullptr)
+    //  linked_module->setPIELevel(PIELevel::Large);
 
 #else
     POCL_MSG_ERR("SPIR not supported\n");

