From 5f6be821cf8c183bba83393b2c32b31aadb1e182 Mon Sep 17 00:00:00 2001
From: Isuru Fernando <isuruf@gmail.com>
Date: Mon, 26 Dec 2022 08:15:05 +0530
Subject: [PATCH] Initialize the command_buffer_khr mutex

---
 lib/CL/clCreateCommandBufferKHR.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/lib/CL/clCreateCommandBufferKHR.c b/lib/CL/clCreateCommandBufferKHR.c
index 58ea26dfa..929f168e2 100644
--- a/lib/CL/clCreateCommandBufferKHR.c
+++ b/lib/CL/clCreateCommandBufferKHR.c
@@ -113,6 +113,7 @@ POname (clCreateCommandBufferKHR) (
       = (cl_command_queue *)calloc (num_queues, sizeof (cl_command_queue));
   memcpy (cmdbuf->queues, queues, num_queues * sizeof (cl_command_queue));
   cmdbuf->num_properties = num_properties;
+  POCL_INIT_LOCK(cmdbuf->mutex);
   cmdbuf->properties = (cl_command_buffer_properties_khr *)calloc (
       num_properties,
       sizeof (cl_command_buffer_properties_khr) * 2 * num_properties + 1);
