diff --git a/lib/CL/devices/cuda/pocl-cuda.c b/lib/CL/devices/cuda/pocl-cuda.c
index df764e0a..d4db5c45 100644
--- a/lib/CL/devices/cuda/pocl-cuda.c
+++ b/lib/CL/devices/cuda/pocl-cuda.c
@@ -1644,7 +1644,7 @@ pocl_cuda_submit_kernel (CUstream stream, _cl_command_node *cmd,
 
   unsigned arg_index = meta->num_args;
 
-  if (meta->num_locals != 0)
+  if (sharedMemBytes != 0)
     {
       /* Deal with automatic local allocations if there are local function args
        */
