diff --git a/CMakeLists.txt b/CMakeLists.txt
index 45b92446..1c61207e 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1825,7 +1825,9 @@ endif()
 ##########################################################
 
 if (ENABLE_LLVM AND NOT CMAKE_CROSSCOMPILING)
-  CHECK_ALIGNOF("double16" "typedef double double16  __attribute__((__ext_vector_type__(16)));" ALIGNOF_DOUBLE16)
+  # CHECK_ALIGNOF("double16" "typedef double double16  __attribute__((__ext_vector_type__(16)));" ALIGNOF_DOUBLE16)
+  message("disabled CHECK_ALIGNOF")
+  set(ALIGNOF_DOUBLE16 128)
 else()
   set(ALIGNOF_DOUBLE16 128)
 endif()
