diff --git a/lib/CL/pocl_file_util.c b/lib/CL/pocl_file_util.c
index 281f6dbf84..377c653ca3 100644
--- a/lib/CL/pocl_file_util.c
+++ b/lib/CL/pocl_file_util.c
@@ -212,24 +212,6 @@ pocl_write_file (const char *path, const char *content, uint64_t count,
   assert(content);
   char path2[POCL_MAX_PATHNAME_LENGTH];
   int err, fd = -1;
-
-  if (pocl_exists(path)) 
-    {
-      if (dont_rewrite) 
-        {
-          if (!append)
-            return 0;
-        } 
-      else 
-        {
-          int res = pocl_remove(path);
-          if (res)
-            {
-              POCL_MSG_ERR ("pocl_remove(%s) failed\n", path);
-              return res;
-            }
-        }
-    }
   
   if (append)
     {
