diff --git a/tests/CMakeLists.txt b/tests/CMakeLists.txt
index 07648a00..e8307056 100644
--- a/tests/CMakeLists.txt
+++ b/tests/CMakeLists.txt
@@ -38,38 +38,6 @@ endif()
 
 #######################################################################
 
-add_test("pocl_test_dlopen_libpocl" "runtime/test_dlopen")
-set_property(TEST "pocl_version_check" APPEND PROPERTY DEPENDS "pocl_test_dlopen_libpocl")
-
-if(BUILD_BASIC)
-  add_test("pocl_test_dlopen_device_basic" "runtime/test_dlopen" "basic")
-  set_property(TEST "pocl_version_check" APPEND PROPERTY DEPENDS "pocl_test_dlopen_device_basic")
-endif()
-
-if(BUILD_PTHREAD)
-  add_test("pocl_test_dlopen_device_pthread" "runtime/test_dlopen" "pthread")
-  set_property(TEST "pocl_version_check" APPEND PROPERTY DEPENDS "pocl_test_dlopen_device_pthread")
-endif()
-
-if(BUILD_ACCEL)
-  add_test("pocl_test_dlopen_device_accel" "runtime/test_dlopen" "accel")
-  set_property(TEST "pocl_version_check" APPEND PROPERTY DEPENDS "pocl_test_dlopen_device_accel")
-endif()
-
-if(ENABLE_TCE)
-  add_test("pocl_test_dlopen_device_tce" "runtime/test_dlopen" "tce")
-  set_property(TEST "pocl_version_check" APPEND PROPERTY DEPENDS "pocl_test_dlopen_device_tce")
-endif()
-
-if(ENABLE_HSA)
-  add_test("pocl_test_dlopen_device_hsa" "runtime/test_dlopen" "hsa")
-  set_property(TEST "pocl_version_check" APPEND PROPERTY DEPENDS "pocl_test_dlopen_device_hsa")
-endif()
-
-if(ENABLE_CUDA)
-  add_test("pocl_test_dlopen_device_cuda" "runtime/test_dlopen" "cuda")
-  set_property(TEST "pocl_version_check" APPEND PROPERTY DEPENDS "pocl_test_dlopen_device_cuda")
-endif()
 
 #######################################################################
 
