From aa5595f8713bb5b5068b439b9a51d586d9fc4979 Mon Sep 17 00:00:00 2001
From: Isuru Fernando <isuruf@gmail.com>
Date: Wed, 17 Jan 2024 00:51:18 -0600
Subject: [PATCH 4/4] Fix typo

---
 pocld/pocld.cc | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/pocld/daemon.cc b/pocld/daemon.cc
index bd674cd148..efb81aa099 100644
--- a/pocld/daemon.cc
+++ b/pocld/daemon.cc
@@ -50,7 +50,7 @@
 #include "daemon.hh"

 #ifndef POLLRDHUP
-#define PULLRDHUP 0
+#define POLLRDHUP 0
 #endif
 #define POLLFD_ERROR_BITS (POLLHUP | POLLERR | POLLNVAL | POLLRDHUP)
 
