From 8d2f28ab2422a23d439a9de0b7af103e403f2943 Mon Sep 17 00:00:00 2001
From: Isuru Fernando <isuruf@gmail.com>
Date: Tue, 12 May 2020 14:58:10 -0500
Subject: [PATCH 1/2] pass --nvptx-short-ptr

---
 lib/CL/devices/cuda/pocl-cuda.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/lib/CL/devices/cuda/pocl-cuda.c b/lib/CL/devices/cuda/pocl-cuda.c
index 16b6dc9e4..38036240a 100644
--- a/lib/CL/devices/cuda/pocl-cuda.c
+++ b/lib/CL/devices/cuda/pocl-cuda.c
@@ -148,6 +148,7 @@ pocl_cuda_init_device_ops (struct pocl_device_ops *ops)
   ops->free_event_data = pocl_cuda_free_event_data;
   ops->join = pocl_cuda_join;
   ops->flush = pocl_cuda_flush;
+  ops->init_build = pocl_cuda_init_build;
   // TODO
   ops->map_mem = pocl_cuda_map_mem;
 
@@ -1570,3 +1571,8 @@ pocl_cuda_finalize_thread (void *data)
 
   return NULL;
 }
+
+char* pocl_cuda_init_build(void *data)
+{
+   return strdup("-mllvm --nvptx-short-ptr");
+}

From 51fc46378cb39c89371cf403d28a3803055d22a4 Mon Sep 17 00:00:00 2001
From: Isuru Fernando <isuruf@gmail.com>
Date: Wed, 13 May 2020 09:01:36 -0500
Subject: [PATCH 2/2] Guard against LLVM<7

---
 lib/CL/devices/cuda/pocl-cuda.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/lib/CL/devices/cuda/pocl-cuda.c b/lib/CL/devices/cuda/pocl-cuda.c
index 38036240a..11314a62a 100644
--- a/lib/CL/devices/cuda/pocl-cuda.c
+++ b/lib/CL/devices/cuda/pocl-cuda.c
@@ -1574,5 +1574,9 @@ pocl_cuda_finalize_thread (void *data)
 
 char* pocl_cuda_init_build(void *data)
 {
-   return strdup("-mllvm --nvptx-short-ptr");
+#ifdef LLVM_OLDER_THAN_7_0
+    return strdup("");
+#else
+    return strdup("-mllvm --nvptx-short-ptr");
+#endif
 }
