diff --git a/pocld/daemon.cc b/pocld/daemon.cc
index bdf711cf8..f8aa50a1c 100644
--- a/pocld/daemon.cc
+++ b/pocld/daemon.cc
@@ -308,18 +308,25 @@ int PoclDaemon::launch(std::string ListenAddress, struct ServerPorts &Ports,
   addrinfo *ai = ResolvedAddress;
   NumListenFds = 0;
   for (addrinfo *ai = ResolvedAddress; ai; ai = ai->ai_next) {
+#ifdef ENABLE_VSOCK
     if (ai->ai_family != AF_INET && ai->ai_family != AF_INET6 &&
         ai->ai_family != AF_VSOCK)
       continue;
+#else
+    if (ai->ai_family != AF_INET && ai->ai_family != AF_INET6)
+      continue;
+#endif
     struct sockaddr *base_addr = ai->ai_addr;
     int base_addrlen = ai->ai_addrlen;
     std::string addr_string = describe_sockaddr(base_addr, base_addrlen);
+#ifdef ENABLE_VSOCK
     if (UseVsock && ai->ai_family != AF_VSOCK) {
       POCL_MSG_ERR("Using vsock requires using the correct address "
                    "vsock:<cid>, instead of %s\n",
                    addr_string.c_str());
       break;
     }
+#endif
     int listen_command_fd = 0;
     int listen_stream_fd = 0;
     struct sockaddr_storage server_addr_command, server_addr_stream;
diff --git a/pocld/shared_cl_context.cc b/pocld/shared_cl_context.cc
index 6bc358bbe..b1678af03 100644
--- a/pocld/shared_cl_context.cc
+++ b/pocld/shared_cl_context.cc
@@ -23,6 +23,7 @@
    IN THE SOFTWARE.
 */
 
+#include <unistd.h>
 #include <cassert>
 #include <cstdio>
 #include <filesystem>
@@ -630,7 +631,7 @@ SharedCLContext::SharedCLContext(cl::Platform *p, unsigned pid,
     if (MaxMemAllocSize < MaxSVMAllocSize)
       MaxSVMAllocSize = MaxMemAllocSize;
 
-    MaxTotalAllocatableSVM = std::min(MaxTotalAllocatableSVM,
+    MaxTotalAllocatableSVM = std::min(MaxTotalAllocatableSVM, (size_t)
                                       Dev.getInfo<CL_DEVICE_GLOBAL_MEM_SIZE>());
   }
 
