diff --git a/tests/runtime/test_dlopen.c b/tests/runtime/test_dlopen.c
index be4bf5eb..cc97fdb8 100644
--- a/tests/runtime/test_dlopen.c
+++ b/tests/runtime/test_dlopen.c
@@ -28,7 +28,13 @@ int
 main (int argc, char **argv)
 {
   int ret = 0;
-  const char *libpocl = "$ORIGIN/../../lib/CL/libpocl.so";
+  const char *libpocl = "$ORIGIN/../../lib/CL/libpocl"
+#ifdef __APPLE__
+          ".dylib";
+#else
+          ".so";
+#endif
+
   char libdevice[4096] = "";
   if (argc > 1)
     snprintf (libdevice, sizeof (libdevice),
